### DISCUSSION TITLE
Mutability consistency

### DISCUSSION THREAD

**Discussion by sebersole** [hibernate/hibernate-orm#6090] [Design Proposals]
Hibernate has different notions of (im)mutability:

1. Whether the internal state of a value can change.  E.g. a Java `Date` is mutable because one can change its internal state via `Date#setTime`.  A `String`, however, is immutable because there is no way to change its internal state.  This sense describes part of the metadata for the Java model and has an affect on dirty check, caching and making copies.  
4. Whether the columns of the attribute are updateable.  In this sense, it is a convenience for `@Column(updatable=true/false)` for all associated columns.  This sense is a function of the O/R mapping.

For the rest of the Discussion, to avoid confusion, let's call (1) mutability and (2) updateability.

Historically Hibernate had `@Immutable`.  It was actually never well defined which of the two `@Immutable` covered; but, as `@Immutable` was only allowed on an entity, this distinction was not super important - an entity we are told is immutable is naturally non-updateable as well since we never update those columns.

In 6.0 we introduced `MutabilityPlan` and `@Mutability` which explicitly deal with (1).  `@Mutability` is allowed on attributes, though not on an entity.

In 6.2 (non-Final atm) we have added the ability to apply `@Immutable` on attributes.  The current effect of that is that the attribute is considered both immutable (1) and non-updateable (2).

The consistency problem is that currently `@Immutable` sometimes means (1), sometimes (2) and sometimes both depending on where it is used.  I saw a few options to achieve this consistency - 

1. In a clean-room, I think what makes the most sense is to consider `@Immutable` a shorthand for an immutable `MutabilityPlan`.  For the notion of updateability I would have either relied on `@Column` or introduced a new annotation (`@Updateable`?) specifically for case (2).  Keeping in mind that allowing `@Immutable` on attributes is not in a Final release, there is no backwards compatibility concern here.  I also like that the names actually reflect the intent.
2. An option we discussed on https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev/topic/.40Immutable.20and.20AttributeConverter[Zulip] was to consider `@Immutable` as always meaning both.  The case of immutable but updateable would have to explicitly use `@Mutability`.  This has a big impact, including some backwards compatibility concerns.

## Where each is valid

Each of `@Immutable` and `@Mutability` (and `@Updateable` if we elect that option) are valid or invalid based on where they are specified:

### Entity

Historically, Hibernate has allowed `@Immutable` on an entity.  I think the effect there is perfectly fine.  `@Mutability` is never appropriate on an entity; Hibernate has very specific, internal requirements for the `MutabilityPlan` for an entity.

With either of the 2 options I listed:

* `@Immutable` would remain valid on the root of an entity hierarchy and continues to mean that Hibernate ignores any attribute state changes to the entities in that hierarchy.  It effectively becomes read only.
* `@Mutability` would remain invalid on an entity

### Attribute

An attribute can be either, neither or both.  It is mutable/immutable independent of being updateable/non-updateable.  

Non-updateable implies immutable.

Generally the mutability of an attribute is dictated by its Java type, as in the `Date` / `String` discussion.   There are times a user might want to make an attribute, with an inherently mutable type (e.g. `Date`), immutable.  E.g. this has some performance gains for the expectation that Hibernate ignores internal state changes. 

A non-updateable attribute means that none of the columns associated with the attribute are considered updateable.  Note that the effective impact of this might vary depending on the type of attribute (basic versus many-to-one e.g.).

Again, this was added in 6.2 which is not yet Final, so we have some leeway here with regard to backwards compatibility.

Currently:

* `@Immutable` is allowed and is shorthand for `@Mutability(Immutability.class)` plus `Column(..., updatable=false)` on all columns associated with the attribute.  
* `@Mutability` is allowed, supplying a custom `MutabilityPlan`

With the clean room option:

* `@Immutable` is allowed and only affects mutability - shorthand for `@Mutability(Immutability.class)`
* `@Mutability` is allowed, supplying a custom `MutabilityPlan`
* `@Updateable(false)` is allowed and is shorthand for `Column(..., updatable=false)` on all columns associated with the attribute

With the Zulip option:

* `@Immutable` is allowed and is shorthand for `@Mutability(Immutability.class)` plus `Column(..., updatable=false)` on all columns associated with the attribute (tbf I only verified basic).  This is how it worked in the previous 6.2 CR releases.
* `@Mutability` is allowed, supplying a custom `MutabilityPlan`
* There is no  `@Updateable`.  `@Updateable(false)` is accomplished with `@Immutable`

See also the discussions below about `AttributeConverter` and `UserType` as they ultimately apply to attributes.  

### AttributeConverter

An `AttributeConverter` can be marked as immutable which implies immutability to every attribute it is applied to.  It is not reasonable for an `AttributeConverter` to be marked as non-updateable.

With the clean room option:

* `@Immutable` is allowed on the converter class and is shorthand for `@Mutability(Immutability.class)` 
* `@Mutability` is allowed, supplying a custom `MutabilityPlan` that is applied to attributes the converter is applied to
* `@Updateable` is not allowed

With Zulip option:

* `@Immutable` is not allowed on the converter class.  
* `@Mutability` is allowed, supplying a custom `MutabilityPlan` that is applied to attributes the converter is applied to

### UserType

Same as `AttributeConverter`

**Comment by yrodiere** [comment]
I may be misunderstanding something, but in my mind immutability *may* imply non-updateability, but not always. In theory, and lacking any input from users, updateability would be a direct consequence of "aggregated" immutability...

What I mean is, for example, with a model like this:

```
public class MyEntity {

    @Embedded
    private MyEmbeddable embedded;

}

public class MyEmbeddable {

    private Date myDate;

}
```

Then if all of the following are considered immutable:

* `MyEntity#embeddable`
* `MyEmbeddable#myDate`
* `Date` (or, if there is a `UserType`/`AttributeConverter`, that type/converter)

... we can safely consider the column(s) bound to `embeddable.myDate` as non-updateable. If any of the above is considered mutable, then we might need to update the column(s) bound to `embeddable.myDate`, so they should be updateable.

Though I'll admit I don't fully grasp what's at stake here. In particular, I don't understand why one would want to force non-updateability on a mutable attribute/type... does that even make sense? Why not declare that attribute/type immutable?

**Reply by sebersole** [reply]
> I may be misunderstanding something, but in my mind immutability _may_ imply non-updateability, but not always.

I never said that.  I said the inverse - non-updateability implies immutability.  To a degree anyway.  I don't think we'd actually want to enforce that - the immutability works as far as dirty-checking goes, but for caching and making copies its still best/safest to treat it asmutable.

> Then if all of the following are considered immutable:

I'm really not understanding what you are trying to get at here.  What makes any of that immutable?

> In particular, I don't understand why one would want to force non-updateability on a mutable attribute/type... does that even make sense? Why not declare that attribute/type immutable?

Mutability leads to Hibernate making a number of choices regarding how to handle those values in certain cases as discussed earlier - dirty-checking, caching and copying.  There are obviously performance considerations there.  Knowing that a value is considered mutable means that Hibernate must:

1. Use `.equals` for comparison, instead of `==` e.g.
2. Make deep copies of the value in numerous situations (L2 cache interactions, state array creation, ...)

Generally, using `.equals` for comparison is not a big deal.  But the deep copies can be and that is exactly where a user might want to tell Hibernate "I'm not planning on changing the internal state for this value" in exchange for better performance.  

It's really evident when an `AttributeConverter` is involved.  By default, aside from a few special cases, we *have to* assume that the converted value is mutable.  The only consistent deep copy algorithm we can rely on is to use the converter itself.  So consider we are loading an entity with a mutable converted value.  When we read that value from the `ResultSet`, we need to call `#convertToEntityAttribute` to get the domain value.  When we read entity state, we also need to create the "loaded state array" which we use later for dirty-checking, etc.  When any of the state is mutable we need to create a deep copy.  In the case of a mutable converted value, that means we need to call `converter.convertToEntityAttribute( converter.convertToDatabaseColumn( loadedDomainValue ) )` for the deep copy.  For just a simple read, we are now up to 3 calls to the converter.  Caching that value?  Well, now we need to do another deep copy to push that value to the cache.  It starts to add up.

Now imagine that the user could tell Hibernate that this converted value is immutable instead.  Now we get out of needing those deep copies.

**Reply by sebersole** [reply]
> I may be misunderstanding something, but in my mind immutability may imply non-updateability, but not always.

In fact, I'd say there really is not correlation in this direction - it is perfectly natural for an immutable attribute (`String`) to be updateable.

**Reply by yrodiere** [reply]
> What makes any of that immutable?

Whatever annotations or configuration you add. I was illustrating how immutability can imply non-updateability in principle, not discussing your specific solution.

> I'm really not understanding what you are trying to get at here. 

It seemed to me that main difference between your two solutions is how immutability and updateability get conflated (or not). So the relationship between the two is what I was discussing.

Specifically, I'm trying to get at "why would we need to specify updateability explicitly when we can infer it from immutability?". Also: "what is the point of marking a property/column as non-updateable if a related property is considered mutable?"

I'm just challenging the idea that we need to have separate concepts at the API level.

> Now imagine that the user could tell Hibernate that this converted value is immutable instead. Now we get out of needing those deep copies.

I understand the benefits of marking something as immutable.

What I don't understand is: 

> why one would want to force non-updateability on a mutable attribute/type

I.e. why "@Updateable(false)" would ever be needed, when:

* `@Immutable` has obvious benefits, and can be used to automatically infer non-updateability.
* The use case for a non-updateable but mutable property is unclear to me.

**Reply by yrodiere** [reply]
> > I may be misunderstanding something, but in my mind immutability may imply non-updateability, but not always.
> 
> In fact, I'd say there really is not correlation in this direction - it is perfectly natural for an immutable attribute (`String`) to be updateable.

It depends.

If the attribute type is immutable and the attribute (Java field) is mutable, sure, it's natural for the column to be updateable.

If the attribute type is immutable and the attribute (Java field) is *also* immutable, it seems more natural for the column to be non-updateable?

**Reply by sebersole** [reply]
> > > I may be misunderstanding something, but in my mind immutability may imply non-updateability, but not always.
> > 
> > 
> > In fact, I'd say there really is not correlation in this direction - it is perfectly natural for an immutable attribute (`String`) to be updateable.
> 
> If the attribute type is immutable and the attribute (Java field) is _also_ immutable, it seems more natural for the column to be non-updateable?

That's not correct.  The mapping you mention would look like:

```
@Immutable
String name;
```

Now, today, that would in fact *force* `name` to be non-updateable.  But that has more to do with a (new) interpretation of `@Immutable` to imply something about updateability.  That also forces `@Immutable` and `@Mutability` to be not correlated - `@Immutable` has implication on updateability while `@Mutability` does not.

But in the clean room option (really the logical point of view), just because this attribute it immutable has no bearing at all over whether it is updateable.

I would guess you are making the misinterpretation that "attribute (Java field) is also immutable" is the same as "attribute (Java field) is non-updateable" (aka `@Immutable` implies non-updateable)

**Reply by yrodiere** [reply]
Ok, so back to my question: what is the purpose of binding an immutable Java field with an immutable type to an updateable column?

Hibernate ORM will never update it, right?

Is it just so that SQL scripts can still update it? If so, wouldn't it make more sense to make that column non-updateable by default, and allow users to make it updateable with an annotation?

**Reply by sebersole** [reply]
> Ok, so back to my question: what is the purpose of binding an immutable Java field with an immutable type to an updateable column?

Assuming we go with the mutability/updateability split, then an example of the mapping you are asking about would be what I posted above:

```
@Immutable
String name;
```

We have an immutable type (`String`).  Here, the `@Immutable` is unnecessary - it's already immutable.  And it is updateable.  I mean that's perfectly valid, fairly normal mapping right?

> Hibernate ORM will never update it, right?

And no, the column here does get updated; it is not mapped as non-updateable.  That would look like either of these:

```
@Column(updatable=false)
String name;
```

or 

```
@Updateable(false)
String name;
```

**Reply by yrodiere** [reply]
> it's already immutable. And it is updateable. I mean that's perfectly valid, fairly normal mapping right?

I'm sorry but no, I don't think that makes much sense. And I still don't see the purpose of that.

To me either a property is (deeply) immutable and then there's no reason for it to be updateable, or it's mutable at least at some level, and then yes, sure, it makes sense for it to be updateable.

But deeply immutable (hence never updated) and updateable... No, I don't understand how that's useful.

**Reply by sebersole** [reply]
Well I think you are still missing the point of mutability.  Maybe I'm wrong though.

```
@Entity class TheEntity {
    ...
    String name;
    Date start;
    @Immutable
    Date end;
}
```


`name` is immutable - we cannot change it's internal state.  But obviously nothing stops us from changing the attribute:

```
TheEntity theEntity = ...;
theEntity.setName( "the new name" );
```

This clearly triggers an update (its updateable).

`start`, as a `Date` is mutable - therefore there are 2 ways to trigger dirtiness:

```
TheEntity theEntity = ...;
theEntity.setStart( ... );
```

and

```
TheEntity theEntity = ...;
theEntity.getStart().setTime( ... );
```

As stated numerous times, because it is mutable we have to use `.equals` to check dirtiness and we have to create a deep copy when creating the TheEntity state array and when interacting with L2 cache.

`end`, as as `Date`, is normally mutable.  However, here we have explicitly told Hibernate to treat it as immutable.  This has a few impacts centered on the fact the we will ignore any internal state change.  So while this dirties the entity:

```
TheEntity theEntity = ...;
theEntity.setEnd( ... );
```

this does not:

```
TheEntity theEntity = ...;
theEntity.getEnd().setTime( ... );
```

The important implication really is that we can skip the deep copies, generally speaking.


> But deeply immutable (hence never updated) and updateable... No, I don't understand how that's useful.

Right, because (I think) you keep confusing mutable and updateable.

**Reply by yrodiere** [reply]
> name is immutable - we cannot change it's internal state. But obviously nothing stops us from changing the attribute:

Ok, that's probably the source of the misunderstanding.

Then my objection would be that the meaning you assign to `@Immutable` is confusing, at least to me. I'd expect this:

```
    @Immutable
    Date end;
```

... to mean that the `end` field won't be mutated (but the value it holds still can, that's unrelated). Because, well, I just annotated a field with `@Immutable`, so that field is not mutable, right?

With *that* meaning, a similar field of type `String` could be considered non-updateable automatically.

Now I guess if your `@Immutable` annotation is annotated with something like `@java.lang.annotation.Target({ElementType.TYPE_USE})`, usage and meaning would be consistent: in the example above, the annotation would apply to `Date`, not to `Date end`. But that's still confusing as hell; probably even more so.

"Do you have anything constructive to suggest instead of wasting my time", you'll say, and you'll be right.

I suppose `@MutabilityOverride` would make sense in this context, and, while verbose, would at least suggest a similar behavior to `@AttributeOverride`/`@AssociationOverride` which affects the type of the attribute rather than the attribute itself. And (at least in the example above) there's no way to interpret this as affecting the field itself, since where would be the override then?

**Reply by sebersole** [reply]
> I suppose @MutabilityOverride would make sense in this context

We already have `@Mutability` to allow users to plug in custom "mutability overrides".

> Then my objection would be that the meaning you assign to @Immutable is confusing

I don't think it really is though.  So long as the meaning is consistent regardless of the placement, that is an easy thing to communicate - `@Immutable` == "immutable internal state".  And it is also consistent with `@Mutability`.

Also, the meaning you ascribe to `@Immutable` is just not at all what it has meant for its entire existence.  6.2 added the ability to apply `@Immutable` on attributes whereas previously that was not allowed.  As part of that work, when `@Immutable` is used on a `@Basic` attribute, it also added this additional non-updateable semantic.  In my opinion, that introduced an inconsistency in the message - the meaning of "immutable" was no longer the same depending on where it was used.

**Comment by sebersole** [comment]
The conclusion after discussions with others on the team is to drop the implication that a `@Immutable` is non-updateable.  This is consistent with what was called "clean room" option.

Does it make sense to provide an `@Updateable` annotation?  I think for the time being I'll skip that part, or at least it is a different discussion.

**Reply by gavinking** [reply]
I dunno about this. To me this runs into the exact same issue that @yrodiere brought up. His intuition (and mine too) is that if I write:

```java
@Immutable
Date end;
```

Then that means that the field `end` cannot be set, that is, that it always refers to the same `Date` instance, and *it is therefore a logical implication* that the column `end` is not updatable.

And I think *every single user* who looks at that is going to think the same thing, no matter what you write in the documentation. Because that's just how people use the word "immutable" with respect to fields. 

Telling them:

> oh no it means the `Date` is immutable, not that the field `end` is immutable

 is going to be an absolutely huge WTF to every single user.

So if you want to go with this more limited semantic for `@Immutable`, then I think you have to actually *disallow it* at the field level, and *only* allow it at the class level. 

Furthermore,  we run into the problem that it's extremely difficult/inconvenient to use `final` on fields of JPA entities, since they must have a default constructor. In the other two options, `@Immutable` was functioning as a substitute for `final`. Here you proposed to introduce `@Updatable(false)` to indicate finality, but I guess I would propose that we simply call this `@Final`, since that's a lot less verbose.

So I'm OK with this only if:

- `@Immutable` is disallowed at the field level, and
- we introduce `@Final` to mark fields which can't change.

(For the record, this doesn't seem better than the solution we outlined on Zulip, though perhaps it's easier to *understand*.)

**Reply by gavinking** [reply]
To be clear, what this ultimately boils down to is that the difference between your "Zulip" and "Clean Room" options is that:

- in the Zulip option, `@Immutable` is disallowed on `AttributeConverter`s, and we don't need a second annotation, but
- in the Clean Room option, `@Immutable` is disallowed on a field, and we need a new annotation to indicate non-updatability of a field.

~If I'm not mistaken, the two solutions are otherwise identical.~

**Reply by gavinking** [reply]
Well, actually, no, perhaps I'm oversimplifying. Because the nice thing about the Zulip proposal was that you could *not* write things which were conceptually inconsistent like:

```java
@Updatable(false) Date date;
```

or

```java
@Updatable(false) Address amount;
```

where `Address` is a mutable class.

If I'm understanding correctly, under your Clean Room option, here we would still dirty-check `Address`, and maintain its snapshot state, *even though the columns are considered non-updatable*, and that is simply unnecessary work?


Whereas with the semantics we outlined on Zulip, I could write:

```java
@Immutable Date date;
```

or

```java
@Immutable Address amount;
```

and that would imply that the mapped columns were non-updatable, *and also we could avoid the unnecessary work of maintaining a snapshot*.

That seems much better to me?

I mean, I can sorta see you arguing that the user could write something like:

```java
@Updatable(false) @Immutable Date date;
```

But this is verbose, not user-friendly, and as I pointed out above, means that essentially every single user is going to misunderstand what `@Immutable` means when applied to a field (at least the first time they see it).

**Reply by gavinking** [reply]
Yeah, OK, so after a few minutes more reflection, I really dislike this solution, compared to the one we talked about on Zulip.

My basic problem is that neither:

    @Updatable(false) Date date;

Nor:

    @Immutable Date date;

Does what I want in the common case where I have an immutable field.

I'm forced to almost always write:

     @Updatable(false) @Immutable Date date;

Yeah, we can make `@Updateable(false)` less verbose by calling it `@Final`, but that still leaves me with:

     @Final @Immutable Date date;

which, usability-wise is inferior to what we had before.

(In fact, reading back over @yrodiere's comments, I think I agree in essence with almost everything he wrote above.)

**Reply by beikov** [reply]
`updatable = false` implies immutable, so there is no need for this:

```
@Updatable(false) @Immutable Date date;
```

I think that allowing `@Immutable` on the field level through https://hibernate.atlassian.net/browse/HHH-15816 was a mistake, as immutability to me is an aspect of a type and not a field. Updatability on the other hand is something that is a property of a field.

The only use case for `@Immutable` on the field level AFAIU is for mutable types that a user does not control but Hibernate provides support for out of the box i.e. `java.util.Date` and `java.util.Calendar`. I think that in those cases, people should just continue to annotate the field with `@Mutability(ImmutableMutabilityPlan.class)` if that helps avoid misunderstandings.

**Reply by gavinking** [reply]
> `updatable = false` implies immutable

I don't see that anywhere in Steve's proposal. He wrote:

> `@Updateable(false)` is allowed and is shorthand for `Column(..., updatable=false)` on all columns associated with the attribute

This does not seem to imply anything about mutability?

**Reply by beikov** [reply]
If we don't update a column, then mutability is irrelevant to Hibernate. We will never copy a value or compare values for dirtiness.

**Reply by gavinking** [reply]
But OK, for the sake of argument, let's follow your logic, and say that 

- `@Updatable(false)` implies `@Immutable` for the target type, and that
- "immutable" is somehow a bad concept for fields. (I have *no clue* why, I've heard people talk about "immutable" fields or even local variables literally thousands of times, but whatever.)

Then, I would propose that we rename `@Updatable(false)` to `@Final`, and we get something I'm *completely fine* with. It's essentially the same thing we had on Zulip, we've just renamed an annotation. (I view this renaming as unnecessary, but perfectly acceptable.)

So this solution would be:

- `@Immutable` only allowed on types (and `AttributeConverter`s), and means `@Mutability(Immutability.class)` and
- `@Final` allowed only on fields, and implies `@Column(updatable=false)` for the mapped columns and `@Mutability(Immutability.class)` for the field type.

This looks just completely fine to me.

(Note that `@Immutable` is *not* allowed at the field level, which circumvents the issue that @yrodiere and I were bothered by.)

**Reply by beikov** [reply]
Yeah, I'm fine with the concept of a `@Final` annotation as a shorthand for `@Column(updatable = false)`.

**Reply by gavinking** [reply]
> If we don't update a column, then mutability is irrelevant to Hibernate. We will never copy a value or compare values for dirtiness.

OK, so you're arguing that `@Updatable(false)` or `@Final` imply `@Mutability(Immutability.class)`. Perfect! You agree with my point!

**Reply by beikov** [reply]
Since we don't make use of the mutability plan when the column is not updatable, I guess that it doesn't matter what we do with the mutability plan then, but sure, forcing the immutable mutability plan works for me.

**Reply by gavinking** [reply]
> we don't make use of the mutability plan when the column is not updatable

I mean, that's an optimization / implementation detail I don't care about at all. What I care about is the user-visible semantics, which it seems we agree on.

**Reply by sebersole** [reply]
I said from the very beginning that this is all about (1) consistency and (2) unfortunate term overload.

Great, fine.  I'll just drop my PR and we can live with the inconsistency and constantly explain how this works.

BTW, `@Immutable`-on-attribute was never documented.  Would be nice if that happened.

**Comment by gavinking** [comment]
Ahah, I think I might have figured out a major reason for us talking past each other on this stuff. Currently, `updatable=false` triggers some hardcoded behavior that is, if I’m not mistaken, *equivalent* to `ImmutableMutabilityPlan`, namely: no need to take a snapshot or dirty-check. (It probably still copies when putting stuff in the second-level cache though.)

And so, at least from the point of view of the user, non-updatability *always* implies immutability, even though there’s no explicit `MutabilityPlan`.

The funny thing is that after much discussion on Zulip, this was essentially the semantic we converged on. Without realizing, it seems to me, that this is the semantic *we already have*.

**Reply by gavinking** [reply]
For example, Steve didn’t like that my impl of `@Immutable` on fields didn’t do anything about mutability plans. But of course it didn’t need to because snapshotting/dirty-checking are already implicitly bypassed!

OMG this shit is so confusing!

**Comment by sebersole** [comment]
I'm also doing nothing else with `@Immutable`.  It is dead to me lol.
